common/main/cpp/native_c4.cc

```diff
@@ -98,43 +99,36 @@ void litecore::jni::jniLog(const char *fmt, ...) {
     va_end(args);
 }
 
-static bool detach(jint getEnvStat) {
-    if (getEnvStat == JNI_EDETACHED) {
-        if (gJVM->DetachCurrentThread() == 0) return true;
-        C4Warn("logCallback(): doRequestClose(): Failed to detach the current thread from a Java VM");
-    }
-    return false;
+
+//-------------------------------------------------------------------------
+// Logging Callback
+//-------------------------------------------------------------------------
+
+static bool detach(jint envState) {
+    return (envState == JNI_EDETACHED) && (JNI_OK == detachJVM("log"));
 }
 
 static void logCallback(C4LogDomain domain, C4LogLevel level, const char *fmt, va_list ignore) {
     if ((cls_C4Log == nullptr) || (m_C4Log_logCallback == nullptr)) {
-        jniLog("Logger: not initialized");
+        jniLog("logger: Not initialized");
         return;
     }
 
     JNIEnv *env = nullptr;
-    jint getEnvStat = gJVM->GetEnv(reinterpret_cast<void **>(&env), JNI_VERSION_1_6);
-
-    if (getEnvStat == JNI_EDETACHED) {
-        if (attachCurrentThread(&env) != 0) {
-            jniLog("Logger: Failed to attach the current thread for logging");
-            return;
-        }
-    } else if (getEnvStat != JNI_OK) {
-        jniLog("Logger: Could not get the environment: %d", getEnvStat);
+    jint envState = attachJVM(&env, "logger");
+    if ((envState != JNI_OK) && (envState != JNI_EDETACHED))
```

Comment:
This is showing up in so many places now that it might even deserve a little inline method or macro, but it's not a big deal.